### ISSUE TITLE
RadioButton widget

### ISSUE THREAD

**Issue by ogoffart** [slint-ui/slint#2732]
A RadioButton is a widget for which only one is active at the same time.
It can be implemented in .slint the same way we implement CheckBox.
But one difficulty is that we need support to have only one RadioButton active. So the compiler must do some transformation on RadioButton and RadioButtonGroup

Compilers looks at parent hierarchy of RadioButton and require a RadioButtonGroup and connect the `RadioButton.checked` with a binding to that group.

API could look like this.

```
component RadioButtonGroup 
{
   in-out property <int> current-id;
   callback selected(id: int);

}

component RadioButton 
{

  in property <string> text;
  out property <bool> checked:  self.id == current-id;
  in property <int> id;  // could be an arbitrary enum or string later

  // internal property ,  link generated by the compiler
  in-out property current-id <=> group.current-id

  // internal clicked callback => { current-id = self.id;  }
}
```

Some design questions:

-  Should RadioButtonGroup inherits HorizontalBox?  No, let the user free to do anything (even horizontal)
-  Should there be an implicit RadioButtonGroup if there is none?   No, otherwise it's too complicated to set the checked property.
- Should current-ids be a template:   `component RadioButtonGroup<T> {  property <T> current-id; }`?  Maybe in the future

**Comment by ogoffart** [comment]
This also raise quastion on how the compiler will find out what is a RadioButton  and how you can implement your own RadioButton in your own style.
Right now, the compiler recognize some Widgets as special like ListView and TabWidget (cf #861)  and we probably should not recognize anything that's called RadioButton.  Maybe we need compiler macro.

I can see how this could be done with propagated properties (cf #2338)


```
component RadioButtonGroup
{
   in-out property @propagated(radio) <int> current-id;
   //...
}

component RadioButton {
  out property <bool> checked:  @propagated(radio).current-id == self.id;
  //...
}
```

Also we'll have to check how this goes with sub-component

```
component MyRadio inherit Rectangle {
       // ...  (no groups)
            RadioButton  { ... }
      // ...
}


component FooBar {

      RadioButtonGroup {
                // ...
                 RadioButton { } // ok, easy
                 MyRadio { } // should that work?
      }
 
}
```

We probably also want to have some groups for things that are not radio button (mutually exclusive buttons)

**Comment by ogoffart** [comment]
@Delphinidae84  As usual with this kind of comments, this issue is currently not on our short-term roadmap. If it's important to you, we encourage you to take a shot at implementing it. We'd be happy to provide mentoring, tips, and review your pull request.
Our roadmap is flexible and responsive to the needs of our users and customers. If this issue is a blocker for you, please let us know by [email](mailto:info@slint.dev) or [chat](https://chat.slint.dev/). Share details about your project, and we'll consider prioritizing it or offering it as a service.

**Comment by J0hannes101** [comment]
If someone is still looking for a simple solution, you can do it manually as follows:
```slint
export component RadioButtonGroup {
    in-out property <bool> is_checkbox_1  <=> checkbox_1.checked;
    in-out property <bool> is_checkbox_2  <=> checkbox_2.checked;
    in-out property <bool> is_checkbox_3  <=> checkbox_3.checked;

    HorizontalBox {
        checkbox_1 := CheckBox {
            checked: true;
            text: "checkbox_1";
            toggled => {
                if !is_checkbox_1 { 
                    is_checkbox_1 = true;
                }
                
                is_checkbox_2 = false;

                is_checkbox_3 = false;
            }
        }

        checkbox_2 := CheckBox {
            text: "checkbox_2";
            toggled => {
                is_checkbox_1 = false;

                if !is_checkbox_2 { 
                    is_checkbox_2 = true;
                }
                
                is_checkbox_3 = false;
            }
        }

        checkbox_3 := CheckBox {
            text: "checkbox_3";
            toggled => {
                is_checkbox_1 = false;
                
                is_checkbox_2 = false;
                
                if !is_checkbox_3 { 
                    is_checkbox_3 = true;
                }
            }
        }
    }
}
```

**Comment by ogoffart** [comment]
Done in 1.17
